Web Inspector: update `update-LegacyInspectorBackendCommands.rb` to support "condition"

dcrousso opened the pull request:
#### ff3a939214b647406650f7f5557ec452a4e2f1a2
<pre>
Web Inspector: update `update-LegacyInspectorBackendCommands.rb` to support &quot;condition&quot;
<a href="https://bugs.webkit.org/show_bug.cgi?id=243106">https://bugs.webkit.org/show_bug.cgi?id=243106</a>

Reviewed by Patrick Angle.

* Source/WebInspectorUI/Scripts/update-LegacyInspectorBackendCommands.rb:
Include the platform in the path for the output `InspectorBackendCommands.js`.

* Source/JavaScriptCore/Scripts/generate-combined-inspector-json.py:
Pass the list of build flags to `generate-combined-inspector-json.py` so that it can save them in a
new `&quot;conditionFlags&quot;` in the `CombinedDomains.json`. This way, build flags that were set for that
version/platform are known, allowing perfect regeneration of the `InspectorBackendCommands.js` on
other builds/platforms.

* Source/JavaScriptCore/inspector/scripts/codegen/models.py:
(Protocol.__init__):
(Protocol.parse_specification):
* Source/JavaScriptCore/inspector/scripts/generate-inspector-protocol-bindings.py:
(generate_from_specification):
Invoke `preprocess.pl` with the `&quot;conditionFlags&quot;` saved in the JSON.

* Source/JavaScriptCore/inspector/scripts/codegen/generator.py:
(Generator.needs_preprocess): Aded.
* Source/JavaScriptCore/inspector/scripts/codegen/generate_js_backend_commands.py:
(JSBackendCommandsGenerator.output_filename):
(JSBackendCommandsGenerator.needs_preprocess): Added.
Add a way for generators to indicate that they need to invoke `preprocess.pl`.

* Source/WebInspectorUI/Scripts/update-LegacyInspectorBackendCommands.rb: Added.
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/iOS/10.3/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/iOS/11.0/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/iOS/11.3/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/iOS/12.0/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/iOS/12.2/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/iOS/13.0/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/iOS/13.4/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/iOS/14.0/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/iOS/14.5/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/iOS/15.0/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/iOS/15.4/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/iOS/16.0/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/macOS/12.0/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/macOS/12.3/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/macOS/13.0/InspectorBackendCommands.js:
* Source/WebInspectorUI/Versions/Inspector-iOS-14.0.json:
* Source/WebInspectorUI/Versions/Inspector-iOS-14.5.json:
* Source/WebInspectorUI/Versions/Inspector-iOS-15.0.json:
* Source/WebInspectorUI/Versions/Inspector-iOS-15.4.json:
* Source/WebInspectorUI/Versions/Inspector-iOS-16.0.json:
* Source/WebInspectorUI/Versions/Inspector-macOS-12.0.json:
* Source/WebInspectorUI/Versions/Inspector-macOS-12.3.json:
* Source/WebInspectorUI/Versions/Inspector-macOS-13.0.json:
Add necessary `&quot;conditionFlags&quot;` for legacy protocol versions.

* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/DerivedSources.make:
* Source/WebInspectorUI/CMakeLists.txt:

Canonical link: <a href="https://commits.webkit.org/252899@main">https://commits.webkit.org/252899@main</a>
</pre>

JosephPecoraro commented on Source/WebInspectorUI/Scripts/update-LegacyInspectorBackendCommands.rb:
```diff
@@ -54,7 +54,7 @@ def all_tasks
     match = File.basename(version_path).match(/^Inspector\-(.*?)\-([^-]+?)\.json$/)
     if match
-      output_path = File.join $web_inspector_protocol_legacy_path, match[2]
+      output_path = File.join $web_inspector_protocol_legacy_path, match[1], match[2]
```
I predicted this many years ago. So happy =)

JosephPecoraro commented on Source/WebInspectorUI/UserInterface/Protocol/Legacy/iOS/10.3/InspectorBackendCommands.js:
```diff
@@ -25,8 +25,7 @@
 
-// DO NOT EDIT THIS FILE. It is automatically generated from Inspector-iOS-10.3.json
-// by the script: Source/JavaScriptCore/inspector/scripts/generate-inspector-protocol-bindings.py
+// DO NOT EDIT THIS FILE. It is automatically generated from Inspector-iOS-10.3.json by the script: Source/JavaScriptCore/inspector/scripts/generate-inspector-protocol-bindings.py
```
When did this change? 🤔

dcrousso replied:
honestly no idea 😅 i wondered the same thing

JosephPecoraro commented on Source/JavaScriptCore/Scripts/generate-combined-inspector-json.py:
```diff
@@ -29,12 +29,12 @@
-if len(sys.argv) < 2:
-    print("usage: %s [json files or directory of json files ...]" % os.path.basename(sys.argv[0]))
+if len(sys.argv) < 3:
+    print("usage: %s [json files or directory of json files ...] 'flags'" % os.path.basename(sys.argv[0]))
```
Might be better to name this `conditional-flags` instead of just `flags` which sound like switches/options.

JosephPecoraro commented on Source/JavaScriptCore/inspector/scripts/codegen/generator.py:
```diff
@@ -52,6 +52,7 @@ def ucfirst(str):
     'canvas-webgl2': 'CanvasWebGL2',  # Recording.Type.canvas-webgl2
     'webgl': 'WebGL',  # Canvas.ContextType.webgl
     'webgl2': 'WebGL2',  # Canvas.ContextType.webgl2
+    'webgpu': 'WebGPU',  # Canvas.ContextType.webgpu
```
Fly by? Seems fine, but weird it was missing?

dcrousso replied:
yeah.  i think this was (accidentally?) removed when the first WebGPU implementation was ripped out.  without this, regenerating the old files results in `"Webgpu"` instead of `"WebGPU"`

patrickangle commented on Source/JavaScriptCore/Scripts/generate-combined-inspector-json.py:
```diff
@@ -66,6 +69,10 @@
+
     print(string.rstrip())
-print("]}")
+print("], \"flags\": \"" + " ".join(used_flags) + "\"}")
```
Other than a bit of extra parsing overhead on the other end, is there a reason to not have this been a real array of strings? (I guess we do need it to be a space-separated string later when we call to perl, but it seems odd that implementation detail would be present in the JSON files)

dcrousso replied:
no real reason other than that at the other end we use it as a raw string, so it just means moving the `"".join(...)`

i personally prefer this approach cause it better matches the way it's actually used (and therefore implicitly implies to future engineers "hey dont use this as an array, pass it as one big thing")

dcrousso committed:
Web Inspector: update `update-LegacyInspectorBackendCommands.rb` to support "condition"
https://bugs.webkit.org/show_bug.cgi?id=243106

Reviewed by Patrick Angle.

* Source/WebInspectorUI/Scripts/update-LegacyInspectorBackendCommands.rb:
Include the platform in the path for the output `InspectorBackendCommands.js`.

* Source/JavaScriptCore/Scripts/generate-combined-inspector-json.py:
Pass the list of build flags to `generate-combined-inspector-json.py` so that it can save them in a
new `"conditionFlags"` in the `CombinedDomains.json`. This way, build flags that were set for that
version/platform are known, allowing perfect regeneration of the `InspectorBackendCommands.js` on
other builds/platforms.

* Source/JavaScriptCore/inspector/scripts/codegen/models.py:
(Protocol.__init__):
(Protocol.parse_specification):
* Source/JavaScriptCore/inspector/scripts/generate-inspector-protocol-bindings.py:
(generate_from_specification):
Invoke `preprocess.pl` with the `"conditionFlags"` saved in the JSON.

* Source/JavaScriptCore/inspector/scripts/codegen/generator.py:
(Generator.needs_preprocess): Aded.
* Source/JavaScriptCore/inspector/scripts/codegen/generate_js_backend_commands.py:
(JSBackendCommandsGenerator.output_filename):
(JSBackendCommandsGenerator.needs_preprocess): Added.
Add a way for generators to indicate that they need to invoke `preprocess.pl`.

* Source/WebInspectorUI/Scripts/update-LegacyInspectorBackendCommands.rb: Added.
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/iOS/10.3/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/iOS/11.0/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/iOS/11.3/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/iOS/12.0/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/iOS/12.2/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/iOS/13.0/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/iOS/13.4/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/iOS/14.0/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/iOS/14.5/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/iOS/15.0/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/iOS/15.4/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/iOS/16.0/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/macOS/12.0/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/macOS/12.3/InspectorBackendCommands.js:
* Source/WebInspectorUI/UserInterface/Protocol/Legacy/macOS/13.0/InspectorBackendCommands.js:
* Source/WebInspectorUI/Versions/Inspector-iOS-14.0.json:
* Source/WebInspectorUI/Versions/Inspector-iOS-14.5.json:
* Source/WebInspectorUI/Versions/Inspector-iOS-15.0.json:
* Source/WebInspectorUI/Versions/Inspector-iOS-15.4.json:
* Source/WebInspectorUI/Versions/Inspector-iOS-16.0.json:
* Source/WebInspectorUI/Versions/Inspector-macOS-12.0.json:
* Source/WebInspectorUI/Versions/Inspector-macOS-12.3.json:
* Source/WebInspectorUI/Versions/Inspector-macOS-13.0.json:
Add necessary `"conditionFlags"` for legacy protocol versions.

* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/DerivedSources.make:
* Source/WebInspectorUI/CMakeLists.txt:

Canonical link: https://commits.webkit.org/252899@main
